Homework Dmitriev

BrainLUX opened the pull request.

BrainLUX committed:
Build 1.0

BrainLUX committed:
Build 1.1

BrainLUX committed:
Build 1.2

BrainLUX committed:
Build 1.3

buttons update + basket wall added

BrainLUX committed:
Build 1.4

basket fixed, player jump reduced

BrainLUX committed:
Build 1.5

Final

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/BackView.java:
```diff
@@ -0,0 +1,77 @@
+import android.graphics.drawable.Drawable;
+import android.util.AttributeSet;
+
+public class BackView extends android.support.v7.widget.AppCompatImageView {
```
Зачем нужен этот класс? Почему не просто ImageView?

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/CustomView.java:
```diff
@@ -0,0 +1,128 @@
+import android.util.AttributeSet;
+
+
+public class CustomView extends android.support.v7.widget.AppCompatImageView {
```
Лучше так не делать. Если нужно что-то нарисовать, то не нужно создавать для этого битам. Рисуй сразу на canvas, это будет быстрее, проще и меньше памяти

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -1,13 +1,226 @@
+    private TextView money;
+    private int score = 0;
+    private BackView backView;
+    boolean scoredThis = false;
```
Лучше явно указывать видимость

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -1,13 +1,226 @@
 
 public class MainActivity extends AppCompatActivity {
 
+    private ImageView mViewTobeFlung;
```
Если используешь венгерская нотацию, тогда используй везде. Или вообще не используй

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -1,13 +1,226 @@
         setContentView(R.layout.activity_main);
+        getSupportActionBar().hide();
+        mViewTobeFlung = findViewById(R.id.iv_translate_fling);
+        final RelativeLayout mMainLayout = findViewById(R.id.main_layout);
```
Тоже самое про название

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -1,13 +1,226 @@
+                return gestureDetector.onTouchEvent(event);
+            }
+        });
+        mMainLayout.getViewTreeObserver().addOnGlobalLayoutListener(
```
А это зачем?

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -1,13 +1,226 @@
+    private void getExtra() {
+        Bundle extras = getIntent().getExtras();
+        if (extras != null) {
+            walls = extras.getBoolean("walls", true);
```
Ключи лучше делать константами

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -1,13 +1,226 @@
+                    playerDirection = 1;
+                }
+                animateBoy(Math.min(distanceInY, 200));
+                final Paint paint = new Paint();
```
paint можно каждый раз не создавать, а инициализировать 1 раз и использовать далее

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -1,13 +1,226 @@
+                ValueAnimator animator = ValueAnimator.ofFloat(0, 1);
+                animator.setDuration(1000);
+                animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
+                    float w = (float) mViewTobeFlung.getWidth() / 2, h = (float) mViewTobeFlung.getHeight() / 2;
```
Лучше это вынести в отдельный класс или хотя бы метод

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -1,13 +1,226 @@
+        }
+
+        @Override
+        public boolean onFling(MotionEvent downEvent, MotionEvent moveEvent, float velocityX, float velocityY) {
```
Стоит хранить все анимации для того, что бы их потом остановить при уничтожении активити

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -0,0 +1,224 @@
+public class MenuActivity extends AppCompatActivity {
+
+    private static final String TAG = MainActivity.class.getName();
+    MediaPlayer mPlayer;
```
Лучше указывать видимость явно

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    private void initValues() {
+        sp = getSharedPreferences("settings", Context.MODE_PRIVATE);
```
Ключи лучше хранить в константах

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -0,0 +1,224 @@
+        balls = new int[BALL_MAX + 1];
+        balls[0] = R.drawable.ball2;
+        balls[1] = R.drawable.ball;
+        balls[2] = R.drawable.ball3;
```
Наверное стоит хранить мячи в какой-то структуре, что бы вручную не забивать каждое значение в массив

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -0,0 +1,224 @@
+            @Override
+            public void onClick(View v) {
+                chosenBall++;
+                if (chosenBall > BALL_MAX)
```
Лучше всегда использовать фигурные скобки

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -0,0 +1,224 @@
+    @Override
+    protected void onPause() {
+        super.onPause();
+        mPlayer.stop();
```
Стоит проверить mPlayer на null

otopba commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,10 +1,98 @@
+    android:background="@drawable/court">
 
+    <ru.ok.technopolis.basketball.CustomView
+        android:id="@+id/view"
```
id стоит делать более уникальными

otopba reviewed (changes requested):
Очень круто выполнено задание!
Спасибо за такой подход к домашней работе. Даже звуки есть, супер!
Если есть желание, давай помогу довести игру до ума и можно выложить в google play

Оставил комментарии к коду

BrainLUX committed:
Build 1.6

Final fixed

BrainLUX commented:
@otopba fixed

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/BackView.java:
```diff
@@ -1,77 +1,49 @@
+@SuppressLint("AppCompatCustomView")
+public class BackView extends ImageView {
+    private Paint paint;
+    private ArrayList<Line> stack;
```
Лучше всегда оперировать на уровне интерфейсов. То есть писать List<Line> stack

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/BackView.java:
```diff
@@ -1,77 +1,49 @@
-            return true;
-        }
+    public void refresh() {
+        stack = new ArrayList<>();
```
Не стоит каждый раз создавать новый список, можно просто чистить старый 
stack.clear();

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/BackView.java:
```diff
@@ -1,77 +1,49 @@
+    @Override
+    protected void onDraw(Canvas canvas) {
+        Log.d("", "onDraw: " + stack.size());
+        for (int i = 0; i < stack.size(); i++) {
```
Тут можно foreach, но это на вкус

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/BackView.java:
```diff
@@ -1,77 +1,49 @@
-    Bitmap bitmap;
+import java.util.ArrayList;
+
+@SuppressLint("AppCompatCustomView")
```
Лучше наследоваться от AppCompatImageView, если уберешь SuppressLint, студия тебе подскажет это

otopba replied:
Кстати, и почему ImageView? Почему не просто View?

BrainLUX replied:
До этого использовал bitmap, забыл поменять.

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/CustomView.java:
```diff
@@ -1,30 +1,27 @@
-        float mid = (float) h / 2;
-        float min = Math.min(w, h);
+    public void drawStar(Canvas canvas, int from) {
+        float mid = (float) canvas.getHeight() / 2;
```
Во время отрисовки нужно брать высоту ни канваса, а высоту самой View canvas.getHeight() -> getHeight() и с шириной тоже самое

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/CustomView.java:
```diff
@@ -36,93 +33,50 @@ public void initStar(int w, int h) {
 
-    public void drawStar(int count) {
-        Paint paint = new Paint();
+        paint = new Paint();
```
Не нужно создавать никаких объектов в onDraw создай Paint и Path заранее

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/Line.java:
```diff
@@ -0,0 +1,31 @@
+package ru.ok.technopolis.basketball;
+
+class Line {
```
Лучше делать public

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -5,84 +5,76 @@
+    private int playerDirection = 1;
+    private Vibrator vibrator;
+    private MediaPlayer mPlayer;
+    private final String WALL_KEY = "walls";
```
private static final

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -5,84 +5,76 @@
-    MediaPlayer mPlayer;
+    private boolean scoredThis = false;
+    private boolean scoredLast;
+    private float x = Float.POSITIVE_INFINITY, y;
```
Так объявлять переменные не стоит. private float x = Float.POSITIVE_INFINITY, y;
Тут сразу и не заметить что в конце еще и `y` есть

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -5,84 +5,76 @@
+    private boolean threw = false;
+    private int playerDirection = 1;
+    private Vibrator vibrator;
+    private MediaPlayer mPlayer;
```
Либо использовать венгерскуя нотацию везде, или нигде. Я за второй вариант)

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -218,9 +207,12 @@ private void score() {
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        animator.end();
```
Лучше  animator.cancel();

otopba replied:
vibrator.cancel();

otopba replied:
mPlayer.release();

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -13,31 +13,33 @@
+    private boolean walls;
+    private boolean vibro;
+    private int chosenBall = 0;
+    private ArrayList<Integer> balls;
```
Лучше работать с интерфейсами. То есть, писать List<Integer>

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -13,31 +13,33 @@
+    private int chosenBall = 0;
+    private ArrayList<Integer> balls;
+    private SharedPreferences sp;
+    private final String SP_NAME = "settings";
```
private static final

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -13,31 +13,33 @@
+    private boolean music;
+    private boolean walls;
+    private boolean vibro;
+    private int chosenBall = 0;
```
По умолчанию и так 0

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -13,31 +13,33 @@
-    final int BALL_MAX = 2;
-    int[] balls;
-    SharedPreferences sp;
+    private MediaPlayer mPlayer;
```
mPlayer

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -13,31 +13,33 @@
+
 public class MenuActivity extends AppCompatActivity {
 
     private static final String TAG = MainActivity.class.getName();
```
Тег от другого класса

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -13,31 +13,33 @@
 import android.widget.Switch;
 
+import java.util.ArrayList;
+
```
Нужно будет освободить ресурсы как в MainActivity  onDestroy

BrainLUX committed:
Build 1.7

Final fixed

BrainLUX commented:
@otopba fixed

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/BackView.java:
```diff
@@ -38,12 +37,13 @@ public void drawLine(float x1, float y1, float x2, float y2) {
     @Override
     protected void onDraw(Canvas canvas) {
         Log.d("", "onDraw: " + stack.size());
```
И логи тоже не стоит писать в onDraw

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/BackView.java:
```diff
@@ -38,12 +37,13 @@ public void drawLine(float x1, float y1, float x2, float y2) {
-            canvas.drawLine(stack.get(i).getX1(), stack.get(i).getY1(), stack.get(i).getX2(), stack.get(i).getY2(), paint);
+        for (Line line : stack) {
+            canvas.drawLine(line.getX1(), line.getY1(), line.getX2(), line.getY2(), paint);
+
```
Лишняя строка

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/CustomView.java:
```diff
@@ -1,33 +1,29 @@
-
-@SuppressLint("AppCompatCustomView")
-public class CustomView extends ImageView {
+public class CustomView extends View {
```
Вообще, можно сделать название View более осмысленным

otopba replied:
И про   canvas.drawText("x" + count, 115, 85, paint); Тут у тебя пиксели 115 и 85, а это значит, что на разных девайсах будет выглядеть по разному. Лучше в конструкторе перевести из dp в px использовать эти значения

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/CustomView.java:
```diff
@@ -1,33 +1,29 @@
-        Paint paint = new Paint();
         paint.setColor(Color.YELLOW);
-        Path path = new Path();
         paint.setStrokeWidth(fat);
```
Все настройки paint, которые не изменяются, лучше делать один раз, а не каждый раз в onDraw

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MainActivity.java:
```diff
@@ -204,15 +206,18 @@ private void score() {
-        animator.end();
+        animator.cancel();
+        vibrator.cancel();
+        if (soundsPlayer != null)
```
Лучше всегда использовать фигурные скобки. И другие переменные тоже можно проверить на null

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -212,22 +210,30 @@ private void setPaintInterface(int mode) {
     @Override
     protected void onPause() {
         super.onPause();
-        if (mPlayer != null)
```
Скобки, проверка на нул других переменных

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/MenuActivity.java:
```diff
@@ -212,22 +210,30 @@ private void setPaintInterface(int mode) {
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        if (musicPlayer != null)
```
Скобки, проверка на нул других переменных

otopba commented:
Работа принята
10 балов

BrainLUX committed:
Build 1.8

Final fixed

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/StarView.java:
```diff
@@ -6,14 +6,17 @@
     private Paint paint;
     private int count;
     private Path path;
+    private float TEXT_X;
```
Названия переменных не пишутся большими буквами. Только у констант

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/StarView.java:
```diff
@@ -44,17 +44,23 @@ public void drawStar(Canvas canvas, int from) {
+        Log.d("", "StarView: " + TEXT_X + " " + TEXT_Y);
+    }
+
+    float fromDptoPx(float dp) {
```
fromDptoPx -> fromDpToPx

otopba commented on app/src/main/java/ru/ok/technopolis/basketball/StarView.java:
```diff
@@ -23,11 +26,8 @@ public void drawStar(Canvas canvas, int from) {
         float fat = min / 17;
```
Все еще остался canvas.getHeight()

BrainLUX committed:
Build 1.9

Final fixed

BrainLUX commented:
@otopba fixed

BrainLUX committed:
Build 1.91

Final fixed

BrainLUX committed:
Build 1.92

Final fixed

BrainLUX committed:
Build 1.94

BrainLUX committed:
Build 1.95

random ball and player pos after shoot
comments

BrainLUX committed:
Build 2.0

simple physic added

BrainLUX committed:
Build 2.01

OOP resolved (FINALLY)

BrainLUX committed:
Build 2.02

physics + oop

BrainLUX committed:
Build 2.02

physics + oop

BrainLUX committed:
Build 2.02

physics + oop

BrainLUX committed:
Build 2.03

game modes + oop

BrainLUX committed:
Build 2.04

new basket model, dynamic radio button and exit button added, small physics and basket position fix

BrainLUX committed:
Build 2.05

physic and ball start postition fix

BrainLUX committed:
Build 2.06

coin added

BrainLUX committed:
Build 2.06

added score table and random breakable walls

BrainLUX committed:
Build 2.07

BrainLUX committed:
Build 2.07 saving returned, added record table functionality

BrainLUX committed:
Here we go again
